### ISSUE TITLE
Feature request: Code quality metrics

### ISSUE THREAD

**Issue by tujger** [sverweij/dependency-cruiser#523]
I've paid a lot of time searching the tool that able to measure code quality metrics such as afferent, efferent, instability etc. They can be pretty obsolete but otherwise they're simple and expressive, and can help analyze the code.

https://kariera.future-processing.pl/blog/object-oriented-metrics-by-robert-martin/

### Current Behavior

Detect imports and exports, find circular dependences and something else.

### Possible Solution

Tool can analyze imports and exports, so it obviously able to calculate those metrics. It could be great to see these metrics on the files. Some extra possible feature - track the metric for each file during the development, maybe from commit to commit.

### Considered alternatives

Not found.

**Comment by sverweij** [comment]
Hi @tujger this is possible. I even have some code lying around for this, somewhere, but wasn't sure there'd be a demand for this, so thanks for raising this issue :-)

There'll be a few things to consider
- The Efferent and Afferent couplings in Uncle Bob's software metrics are defined in terms of _classes_ within a _component_. The metrics dependency-cruiser can produce will be of (javascript/ typescript) _modules_  within _folders_. This can be similar (and still useful) but it's not 100% the same/ comparable to the same metrics in java or smalltalk.
- As dependency-cruiser works on modules and not classes, it's not possible to calculate the _Abstractness_ and _Distance from the main sequence_ metrics; these require knowledge about whether classes (which might or might not be defined in the module) is abstract or not.
- _Package dependency cycles_ - dependency-cruiser is able to detect (and flag) cyclic dependencies between modules - detecting the same between _components_ (/ _folders_?) requires implementing #324 

Question: what type of output would be useful for your use case? Graphical? Text? [shareable html](https://github.com/sverweij/dependency-cruiser/blob/develop/doc/cli.md#err-html)? Something else?

**Comment by tujger** [comment]
Hi @sverweij,

Great to know it. I like how the tool detects all dependencies between components, so I guess all metrics can be calculated using the same approach. Isnt' it?

**Comment by tujger** [comment]
> Question: what type of output would be useful for your use case? Graphical? Text? [shareable html](https://github.com/sverweij/dependency-cruiser/blob/develop/doc/cli.md#err-html)? Something else?

I guess all these types can be useful. Graphically, it can be shown nearby the module and enhance the analyzing of structure. Text (as a table?) can provide the summary over the folders/files.

**Comment by tujger** [comment]
And let me share a couple of npm scripts I use. It works on Mac, and dot considers the installed Graphviz:

    "scripts": {
      "deps high": "depcruise --config .dependency-cruiser.js src/ --output-type archi > archi.dot && dot -Tpng archi.dot -o archi.png && open archi.png",
      "deps folder": "depcruise --config .dependency-cruiser.js src/ --output-type ddot > archi.dot && dot -Tpng archi.dot -o archi.png && open archi.png",
      "deps detailed": "depcruise --config .dependency-cruiser.js src/ --output-type dot > archi.dot && dot -Tpng archi.dot -o archi.png && open archi.png"
    }

**Comment by sverweij** [comment]
Hi @tujger to test the waters and understand the problem domain I've hacked together a proof of concept [reporter plugin](https://gist.github.com/sverweij/06487b592b39c2e9b4cd9e4ff0a6dc08) that emits efferent & afferent couplings and the instability metric. I'll use what I learn here to incorporate into dependency-cruiser itself (so these metrics are available to any reporter).

If you want to you can already use the code from the gist ([instructions included](https://gist.github.com/sverweij/06487b592b39c2e9b4cd9e4ff0a6dc08#file-_runme-md)).

Would an output such as this be of use? What would need to change in the output?

What I bumped into:

- taking into account external dependencies
  for the proof of concept I've excluded external dependencies from the calculations - this is just to make visually validating the results easier. With the proper implementation this will be different (and may need to become configurable).
- handling nested folders
  `src/report/dot/index.js` is part of the metrics of 3 folders: `src`, `src/report` and `src/report/dot`
- handling folders with no incoming or outgoing dependencies
  `Ce/(Ca + Ce)` will give `NaN` or `infinity` - but in the spirit of Uncle Bob this should probably be 'maximum stable' or 0 - code includes that
- sorting the table
  I've sorted the table on _instability_ - as that seems to give the most information 
- it seemed useful to also mention the (aggregated) amount of modules in the folder
  so that's what's in the _N_ column.

```
folder                                     N    Ca    Ce  I
--------------------------------------- ---- ----- ----- -----
bin                                        4     0    10  1   
src/report/dot                             8     1     8  0.89
src/report                                28     2    11  0.85
src/enrich                                19     2    10  0.83
src/cli/init-config                       10     1     5  0.83
src/main                                  10     4    15  0.79
src/extract                               40     4    15  0.79
src/enrich/derive/reachable                2     1     3  0.75
src/cli                                   24     7    14  0.67
src/extract/transpile                     10     4     8  0.67
src/config-utl/extract-depcruise-config    3     1     2  0.67
src/main/rule-set                          2     2     3  0.6 
src/enrich/derive/circular                 2     1     1  0.5 
src/enrich/derive/dependents               2     1     1  0.5 
src/enrich/derive/orphan                   2     1     1  0.5 
src/extract/resolve                       10     6     6  0.5 
src/report/error-html                      3     1     1  0.5 
src/report/html                            2     1     1  0.5 
src/main/resolve-options                   1     2     2  0.5 
src/cli/listeners                          4     2     2  0.5 
src/cli/listeners/cli-feedback             1     1     1  0.5 
src/cli/listeners/performance-log          3     1     1  0.5 
src/config-utl                             8     5     4  0.44
src/enrich/summarize                       5     3     2  0.4 
src/main/options                           3     3     2  0.4 
src/extract/parse                          3     7     4  0.36
src/enrich/derive                          9     5     2  0.29
src/extract/ast-extractors                 7     5     2  0.29
src/validate                               7     3     1  0.25
src                                      154    10     0  0   
src/graph-utl                             10    18     0  0   
src/utl                                    4    14     0  0   
src/extract/utl                            6    10     0  0   
src/extract/resolve/get-manifest           2     2     0  0   
src/schema                                 3     2     0  0   
src/main/files-and-dirs                    1     1     0  0   
src/main/utl                               1     2     0  0   
src/report/anon                            4     1     0  0   
src/report/utl                             1     2     0  0   
src/report/plugins                         1     1     0  0   
src/cli/utl                                2     4     0  0   
src/cli/tools                              2     1     0  0   
src/cli/tools/svg-in-html-snippets         1     0     0  0 
```

**Comment by tujger** [comment]
It is great! It obviously provides a lot of useful information that I can use to improve structure of my app. Thank you for your work!
And I have a couple of wishes:
1. It's conveniently calculating metrics for folders. And how can I switch to detailed metrics - for all modules specifically? May be some optional key to enable listing all js/tsx files?
2. Now it uses external dependencies for Ca/Ce. How can i calculate it with only imports within my application?
3. May be very useful to watch changes between commits. Or at least between runs. For example, it saves the metrics into file, and next time it shows diffs between, i.e.:

![image](https://user-images.githubusercontent.com/12112326/141792936-232767e3-70c3-4958-af6d-62a144657fc8.png)

**Comment by sverweij** [comment]
Hi @tujger - thanks for the feedback!

> 1. It's conveniently calculating metrics for folders. And how can I switch to detailed metrics - for all modules specifically? May be some optional key to enable listing all js/tsx files?

Uncle Bob's book talks about these metrics in terms of _components_ - a collection of classes. I guess you could see a class as a component of 1 classes (or in js lingo: a module as a folder of 1 modules) with that whether the _stable dependencies principle (SDP)_ for relations between that class (/ that module) and other classes still hold . Maybe on a different level the _SDP_ was envisioned for (?) but that can indeed be useful.

`Ca` & `Ce` for these are easy to calculate (`.dependents.length`, `.dependencies.length`) and as a consequence so would the `instability`. 🤔 

> 2. Now it uses external dependencies for Ca/Ce. How can i calculate it with only imports within my application?

Tell  dependency-cruiser to only run against the sources e.g. if all your sources reside in `src`:

```
depcruise src --config .dependency-cruiser.js --include-only "^src" --output-type metrics
```

This will not only disregard stuff in node_modules, but also node builtins.


> 3. May be very useful to watch changes between commits. Or at least between runs. For example, it saves the metrics into file, and next time it shows diffs between, i.e.:
> 

I can see how that could be useful (e.g. as automated run with PR reviews, just like with coverage reports). I'll leave that as a future feature now, though.

**Comment by tujger** [comment]
React components are frequently incapsulated within one module, so this makes sense to optionally calculate metrics for each file, I guess.

One thing I noted - my imports are absolute in context of ./src. Here is the way to do it: [link](https://dev.to/mr_frontend/absolute-imports-in-create-react-app-3ge8). So, imports look like:

    import Input from "elements/Input"; // internal import
    import {atom, selector} from "recoil"; // external import

Unfortunately, depcruiser doesn't understand it and considers those imports as external. Hence, all metrics are getting incorrect.

Anyway, this tool show me much more useful information than I could imagine. Thank you very much!

**Comment by sverweij** [comment]
Hi @tujger dependency-cruiser should recognize aliases like `elements` for `src/elements` and it should recognize them as internal dependencies.

 Does your .dependency-cruiser.js tell dependency-cruiser where it can find the tsconfig.json (or jsconfig.json) that define these aliases? You can do this with the [tsConfig](https://github.com/sverweij/dependency-cruiser/blob/develop/doc/options-reference.md#tsconfig-use-a-typescript-configuration-file-project) option.

If this doesn't work let me know; you then might have hit a bug that needs fixing.

**Comment by tujger** [comment]
Yes, it has this option. Then, let me explain the cases.

Folder contains 4 files, there are imports sections:

File 1 (1 internal):

    import {setRecoil} from "elements/RecoilUtils"; // internal
    import {getRecoil} from "elements/RecoilUtils"; // internal
    import {useRecoilValue} from "recoil";
    import {useRecoilState} from "recoil";
    import {atom} from "recoil";

File 2 (1 internal):

    import {setRecoil} from "elements/RecoilUtils"; // internal
    import {atomFamily} from "recoil";
    import {useRecoilValue} from "recoil";

File 3: 

    import {atom} from "recoil";

File 4 (3 internals): 

    import {getRecoil} from "elements/RecoilUtils"; // internal
    import {setRecoil} from "elements/RecoilUtils"; // internal
    import {useRecoilState, useRecoilValue} from "recoil";
    import {selectorFamily} from "recoil";
    import {atom, atomFamily} from "recoil";
    import current from "helpers/current"; // internal
    import {findSchema} from "helpers/schemasMethods"; // internal

Finally, folder has 3 internal imports.

Now, the cases.

Calling `depcruise src/ --include-only "^src" --output-type metrics`, got:

    N    Ca    Ce    I
    4    5     0     0

Calling `depcruise src/ --include-only "^src" --output-type metrics --ts-config tsconfig.json`, got:

    N    Ca    Ce    I
    4    40    5     0.11

Calling `depcruise --config .dependency-cruiser.js src/ --include-only "^src" --output-type metrics` (with tsConfig option), got:

    N    Ca    Ce   I
    4    40    5    0.11

**Comment by sverweij** [comment]
hi @tujger I've tried to reproduce the case of the mysteriously-the-same metrics but wasn't able to. For me running with `--include-only "^src"` did yield different results. You can check the reproduction attempt in [dependency-cruiser-repro-repo/523](https://github.com/sverweij/dependency-cruiser-repro-repo/tree/main/523). This just means that there's a difference between the setup presented there and the setup you used. Could you provide a little more details and/ or a minimal reproduction example (you could use the linked folder above as a starting point if that helps)?

**Comment by sverweij** [comment]
B.t.w. I've added module level metrics to the `metrics` reporter as well. Published on npm as dependency-cruiser@10.8.0-beta-1 (shasum: 462710a3ffe685ed5c2f5d111112474b7cf01685). Usage same as before. Typical output looks like:


<details>
<summary>(click to expand)</summary>

command used:
```
depcruise src --config  --output-type metrics
```

With [this .dependency-cruiser.json](https://github.com/sverweij/dependency-cruiser/blob/develop/.dependency-cruiser.json) as configuration.

```
name                                                         N    Ca    Ce  I    
-------------------------------------------------------- ----- ----- -----  -----
src                                                        162     0   222  1   
src/cli                                                     23     0    50  1   
src/cli/tools                                                1     0     2  1   
src/cli/format.js                                            1     0     5  1   
src/cli/index.js                                             1     0    17  1   
src/cli/tools/wrap-stream-in-html.js                         1     0     2  1   
src/cli/validate-node-environment.js                         1     0     2  1   
src/extract                                                 40     4    90  0.96
src/report                                                  29     2    42  0.95
src/cli/init-config                                         10     1    20  0.95
src/report/dot                                               8     1    18  0.95
src/extract/get-dependencies.js                              1     1    15  0.94
src/enrich                                                  24     2    25  0.93
src/config-utl/extract-depcruise-config                      3     1    11  0.92
src/enrich/enrich-modules.js                                 1     1    11  0.92
src/extract/resolve/index.js                                 1     1    11  0.92
src/main                                                    10     4    37  0.9 
src/report/dot/index.js                                      1     1     9  0.9 
src/config-utl/extract-babel-config.js                       1     1     9  0.9 
src/main/report-wrap.js                                      1     1     8  0.89
src/extract/clear-caches.js                                  1     1     7  0.88
src/extract/gather-initial-sources.js                        1     1     7  0.88
src/report/index.js                                          1     2    14  0.88
src/cli/init-config/index.js                                 1     1     7  0.88
src/cli/init-config/write-run-scripts-to-manifest.js         1     1     7  0.88
src/extract/transpile                                       10     4    27  0.87
src/enrich/derive/reachable                                  2     1     6  0.86
src/report/error-html                                        3     1     6  0.86
src/enrich/derive/reachable/index.js                         1     1     6  0.86
src/extract/index.js                                         1     1     6  0.86
src/extract/transpile/vue-template-wrap.js                   1     1     6  0.86
src/config-utl/extract-depcruise-config/index.js             1     1     6  0.86
src/config-utl                                               8     5    27  0.84
src/extract/resolve                                         10     6    30  0.83
src/report/anon                                              4     1     5  0.83
src/extract/resolve/resolve-cjs.js                           1     1     5  0.83
src/report/dot/prepare-custom-level.js                       1     1     5  0.83
src/report/dot/prepare-folder-level.js                       1     1     5  0.83
src/main/rule-set/validate.js                                1     1     5  0.83
src/config-utl/extract-ts-config.js                          1     1     5  0.83
src/main/index.js                                            1     3    13  0.81
src/main/resolve-options                                     1     2     8  0.8 
src/enrich/index.js                                          1     1     4  0.8 
src/validate/index.js                                        1     1     4  0.8 
src/validate/match-dependency-rule.js                        1     1     4  0.8 
src/validate/match-module-rule.js                            1     1     4  0.8 
src/extract/parse/to-javascript-ast.js                       1     2     8  0.8 
src/extract/transpile/coffeescript-wrap.js                   1     1     4  0.8 
src/extract/transpile/svelte-wrap.js                         1     1     4  0.8 
src/extract/transpile/typescript-wrap.js                     1     1     4  0.8 
src/extract/resolve/determine-dependency-types.js            1     1     4  0.8 
src/main/options/normalize.js                                1     1     4  0.8 
src/report/anon/index.js                                     1     1     4  0.8 
src/report/error-html/index.js                               1     1     4  0.8 
src/main/resolve-options/normalize.js                        1     2     8  0.8 
src/cli/normalize-cli-options.js                             1     2     8  0.8 
src/config-utl/extract-depcruise-config/merge-configs.js     1     1     4  0.8 
src/cli/init-config/write-config.js                          1     1     4  0.8 
src/main/rule-set                                            2     2     7  0.78
src/enrich/summarize                                         5     3     9  0.75
src/extract/resolve/get-manifest                             2     2     6  0.75
src/report/html                                              2     1     3  0.75
src/cli/listeners/cli-feedback                               1     1     3  0.75
src/cli/format-meta-info.js                                  1     1     3  0.75
src/enrich/derive/metrics/get-stability-metrics.js           1     1     3  0.75
src/enrich/soften-known-violations.js                        1     1     3  0.75
src/enrich/summarize/index.js                                1     1     3  0.75
src/enrich/summarize/summarize-modules.js                    1     2     6  0.75
src/extract/transpile/babel-wrap.js                          1     1     3  0.75
src/extract/transpile/livescript-wrap.js                     1     1     3  0.75
src/extract/resolve/get-manifest/merge-manifests.js          1     1     3  0.75
src/extract/ast-extractors/extract-amd-deps.js               1     1     3  0.75
src/extract/ast-extractors/extract-es6-deps.js               1     1     3  0.75
src/extract/ast-extractors/swc-dependency-visitor.js         1     1     3  0.75
src/extract/ast-extractors/extract-typescript-deps.js        1     1     3  0.75
src/report/dot/prepare-flat-level.js                         1     1     3  0.75
src/graph-utl/consolidate-to-folder.js                       1     1     3  0.75
src/report/error-html/utl.js                                 1     1     3  0.75
src/report/html/index.js                                     1     1     3  0.75
src/config-utl/extract-depcruise-config/read-config.js       1     1     3  0.75
src/config-utl/extract-known-violations.js                   1     1     3  0.75
src/cli/init-config/build-config.js                          1     1     3  0.75
src/cli/init-config/get-user-input.js                        1     1     3  0.75
src/cli/init-config/normalize-init-options.js                1     1     3  0.75
src/cli/listeners/cli-feedback/index.js                      1     1     3  0.75
src/cli/listeners                                            4     2     5  0.71
src/extract/transpile/meta.js                                1     4    10  0.71
src/extract/parse/to-typescript-ast.js                       1     2     5  0.71
src/extract/resolve/resolve-amd.js                           1     2     5  0.71
src/graph-utl/consolidate-module-dependencies.js             1     2     5  0.71
src/graph-utl/consolidate-modules.js                         1     2     5  0.71
src/report/error.js                                          1     2     5  0.71
src/extract/parse                                            3     7    17  0.71
src/main/options                                             3     3     7  0.7 
src/extract/ast-extractors                                   7     5    11  0.69
src/enrich/derive/dependents                                 2     1     2  0.67
src/cli/listeners/performance-log                            3     1     2  0.67
src/enrich/derive/metrics/folder.js                          1     1     2  0.67
src/validate/violates-required-rule.js                       1     1     2  0.67
src/enrich/derive/dependents/index.js                        1     1     2  0.67
src/enrich/summarize/add-rule-set-used.js                    1     1     2  0.67
src/enrich/summarize/summarize-options.js                    1     1     2  0.67
src/extract/resolve/external-module-helpers.js               1     3     6  0.67
src/extract/resolve/get-manifest/index.js                    1     2     4  0.67
src/extract/resolve/resolve-helpers.js                       1     1     2  0.67
src/main/options/validate.js                                 1     2     4  0.67
src/report/anon/anonymize-path-element.js                    1     1     2  0.67
src/report/dot/theming.js                                    1     2     4  0.67
src/report/metrics.js                                        1     1     2  0.67
src/report/teamcity.js                                       1     1     2  0.67
src/main/rule-set/normalize.js                               1     1     2  0.67
src/cli/init-config/inquirer-validators.js                   1     1     2  0.67
src/cli/listeners/performance-log/index.js                   1     1     2  0.67
src/cli/listeners/performance-log/handlers.js                1     1     2  0.67
src/validate                                                 7     3     5  0.63
src/main/utl                                                 1     2     3  0.6 
src/main/utl/normalize-re-properties.js                      1     2     3  0.6 
src/graph-utl/filterbank.js                                  1     2     3  0.6 
src/extract/parse/to-swc-ast.js                              1     3     4  0.57
src/enrich/derive                                           14     7     9  0.56
src/enrich/derive/metrics                                    5     2     2  0.5 
src/enrich/derive/circular                                   2     1     1  0.5 
src/enrich/derive/orphan                                     2     1     1  0.5 
src/main/files-and-dirs                                      1     1     1  0.5 
src/report/plugins                                           1     1     1  0.5 
src/enrich/clear-caches.js                                   1     1     1  0.5 
src/enrich/derive/metrics/module-utl.js                      1     1     1  0.5 
src/graph-utl/add-focus.js                                   1     2     2  0.5 
src/enrich/add-validations.js                                1     1     1  0.5 
src/enrich/derive/circular/index.js                          1     1     1  0.5 
src/enrich/derive/circular/get-cycle.js                      1     1     1  0.5 
src/enrich/derive/dependents/get-dependents.js               1     1     1  0.5 
src/enrich/derive/metrics/module.js                          1     1     1  0.5 
src/enrich/derive/orphan/index.js                            1     1     1  0.5 
src/enrich/derive/orphan/is-orphan.js                        1     1     1  0.5 
src/enrich/derive/reachable/get-path.js                      1     1     1  0.5 
src/extract/transpile/index.js                               1     1     1  0.5 
src/extract/transpile/svelte-preprocess.js                   1     1     1  0.5 
src/extract/ast-extractors/extract-cjs-deps.js               1     2     2  0.5 
src/extract/ast-extractors/extract-swc-deps.js               1     1     1  0.5 
src/extract/utl/detect-pre-compilation-ness.js               1     1     1  0.5 
src/main/files-and-dirs/normalize.js                         1     1     1  0.5 
src/report/anon/anonymize-path.js                            1     1     1  0.5 
src/report/anon/random-string.js                             1     1     1  0.5 
src/report/csv.js                                            1     1     1  0.5 
src/report/dot/dot.template.js                               1     1     1  0.5 
src/graph-utl/consolidate-to-pattern.js                      1     2     2  0.5 
src/utl/wrap-and-indent.js                                   1     2     2  0.5 
src/report/error-html/error-html.template.js                 1     1     1  0.5 
src/report/error-long.js                                     1     1     1  0.5 
src/report/html/html.template.js                             1     1     1  0.5 
src/report/plugins/index.js                                  1     1     1  0.5 
src/report/text.js                                           1     1     1  0.5 
src/config-utl/extract-webpack-resolve-config.js             1     1     1  0.5 
src/cli/init-config/config.js.template.js                    1     1     1  0.5 
src/cli/init-config/environment-helpers.js                   1     6     5  0.45
src/extract/resolve/resolve.js                               1     4     3  0.43
src/graph-utl                                               10    18    13  0.42
src/graph-utl/rule-set.js                                    1     3     2  0.4 
src/cli/utl                                                  2     4     2  0.33
src/validate/is-module-only-rule.js                          1     2     1  0.33
src/extract/utl/get-extension.js                             1     2     1  0.33
src/report/dot/module-utl.js                                 1     4     2  0.33
src/cli/utl/io.js                                            1     2     1  0.33
src/cli/utl/validate-file-existence.js                       1     2     1  0.33
src/extract/resolve/module-classifiers.js                    1     5     2  0.29
src/extract/ast-extractors/estree-helpers.js                 1     3     1  0.25
src/config-utl/make-absolute.js                              1     3     1  0.25
src/validate/matchers.js                                     1     4     1  0.2 
src/utl                                                      4    14     3  0.18
src/extract/utl                                              6    10     2  0.17
src/enrich/derive/utl.js                                     1     5     1  0.17
src/utl/bus.js                                               1     5     1  0.17
src/extract/utl/path-to-posix.js                             1     5     1  0.17
src/schema                                                   6     2     0  0   
src/report/utl                                               1     2     0  0   
src/cli/defaults.js                                          1     4     0  0   
src/enrich/derive/metrics/utl.js                             1     3     0  0   
src/graph-utl/match-facade.js                                1     3     0  0   
src/utl/bus-log-levels.js                                    1     5     0  0   
src/utl/array-util.js                                        1     2     0  0   
src/validate/utl.js                                          1     4     0  0   
src/enrich/summarize/is-same-violation.js                    1     2     0  0   
src/graph-utl/compare.js                                     1     7     0  0   
src/meta.js                                                  1    16     0  0   
src/extract/transpile/javascript-wrap.js                     1     1     0  0   
src/extract/utl/strip-query-parameters.js                    1     1     0  0   
src/extract/utl/compare.js                                   1     1     0  0   
src/extract/utl/extract-module-attributes.js                 1     1     0  0   
src/schema/cruise-result.schema.js                           1     1     0  0   
src/main/options/defaults.js                                 1     1     0  0   
src/report/baseline.js                                       1     1     0  0   
src/report/utl/dependency-to-incidence-transformer.js        1     2     0  0   
src/report/dot/default-theme.js                              1     1     0  0   
src/graph-utl/strip-self-transitions.js                      1     3     0  0   
src/report/identity.js                                       1     1     0  0   
src/report/json.js                                           1     1     0  0   
src/schema/configuration.schema.js                           1     1     0  0   
src/cli/init-config/utl.js                                   1     2     0  0   
src/cli/listeners/performance-log/format-helpers.js          1     1     0  0   
src/schema/baseline-violations.schema.js                     1     0     0  0   
src/schema/baseline-violations.schema.json                   1     0     0  0   
src/schema/configuration.schema.json                         1     0     0  0   
src/schema/cruise-result.schema.json                         1     0     0  0  
```

</details>

**Comment by tujger** [comment]
Oh, it's really useful! Thank you so much! It works and shows metrics for modules. Just a small question - is it optional? How can I switch between folders/details? :) And how can I change sort type for the table?

About the repo.

Reproduced [here](https://github.com/tujger/dependency-cruiser-repro-repo/tree/main/523)

Shortly:

    $ depcruise src --include-only '^src' --config --output-type metrics
    name                              N    Ca    Ce  I
    ----------------------------- ----- ----- -----  -----
    src/atoms                         5     0     5  1   
    src                               8     0     0  0   
    src/elements                      1     4     0  0   
    src/helpers                       2     1     0  0   

I think I know. For the folders it just summarize metrics for files, even if imports are the same in the different ones. But in fact, folder atoms integrally has only 3 imports despite every file has different imports (8 totally).

**Comment by sverweij** [comment]
hi @tujger I was planning to make the metrics reporter configurable (ability to hide either folders or modules, custom sorting) in a later stage, but tonight needed something to calm my mind and wrote it anyway. With dependency-cruiser@10.8.0-beta-3 you can specify how you want the metrics reporter to behave in the options section of .dependency-cruiser.js e.g.:

```javascript
module.exports = {
  forbidden: [
    // ... your rules
  ],
  options: {
    // ... other options
    reporterOptions: {
      metrics: {
        hideModules: true, // hides the modules from the metrics reporter output
        // hideFolders: true, // would hide folders from the metrics reporter output
        orderBy: 'name', // possible values: name, moduleCount, afferentCouplings, efferentCouplings, instability
      }
    }
  }
}
```

Thanks for the reproduction repo! I'll look into what's going on there later on.

**Comment by tujger** [comment]
Hi
Excellent! It's getting almost as I expected. Thank you very much! :)

**Comment by sverweij** [comment]
hi @tujger been looking at the repro repo. `atoms/file5.ts` is references `helpers/uischemasMethods` - but that doesn't resolve (also visible in the diagrams); likely a typo as `helpers/schemasMethods` _does_ resolve. As the non-resolved _helpers/uischemasMethods_ doesn't match the `^src` pattern it's not counted as belonging in src. Taking that into account the the Ca & Ce numbers seem right to me.

I've run the scripts in the repro repo after correcting the typo in `atoms/file5.ts` and manually counted the dependents and dependencies for all folders and modules, both with  --include-only and without, and they seem to add up as I'd expect them to. 

=> Did I miss something?

#### including external dependencies

<img width="1204" alt="image" src="https://user-images.githubusercontent.com/4822597/143683187-83762ee1-76d2-47e8-8135-2f0f3db9e9d5.png">

#### including only ^src

<img width="1204" alt="image" src="https://user-images.githubusercontent.com/4822597/143683309-96d603b1-1df2-4741-83ea-a750616f7795.png">

**Comment by sverweij** [comment]
@tujger I've published dependency-cruiser@10.8.0 and dependency-cruiser@11.0.0, both of which include the metrics reporter as you've tested it. Inclusion in other reporters and additional checks (e.g. the stable dependency principle) will follow in later release.

Thanks again for this very good suggestion and for the feedback!
